google/cloud/internal/win32/parse_service_account_p12_file.cc

```diff
@@ -0,0 +1,318 @@
+  // Encode the blob to PKCS#1 format.
+  DWORD pkcs1_encoded_length;
+  if (!CryptEncodeObjectEx(X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+                           PKCS_RSA_PRIVATE_KEY, exported_key.data(), 0,
+                           nullptr, nullptr, &pkcs1_encoded_length)) {
+    return InvalidArgumentError(
+        FormatWin32Errors("Could not encode private key from PKCS#12 file (",
+                          source, "): "),
+        GCP_ERROR_INFO());
+  }
+  std::vector<BYTE> pkcs1_encoded(pkcs1_encoded_length);
+  CryptEncodeObjectEx(X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+                      PKCS_RSA_PRIVATE_KEY, exported_key.data(), 0, nullptr,
+                      pkcs1_encoded.data(), &pkcs1_encoded_length);
+  return pkcs1_encoded;
+}
+
+StatusOr<std::vector<BYTE>> EncodeRsaPkcs8PrivateKey(
+    absl::Span<BYTE const> pkcs1_encoded, std::string const& source) {
+  CRYPT_PRIVATE_KEY_INFO private_key_info;
+  private_key_info.Version = 0;
+  private_key_info.Algorithm.pszObjId = szOID_RSA_RSA;
+  private_key_info.Algorithm.Parameters.cbData = 0;
+  private_key_info.Algorithm.Parameters.pbData = nullptr;
+  private_key_info.PrivateKey.cbData = static_cast<DWORD>(pkcs1_encoded.size());
+  private_key_info.PrivateKey.pbData = const_cast<BYTE*>(pkcs1_encoded.data());
+  private_key_info.pAttributes = nullptr;
+  DWORD pkcs8_encoded_length;
+  if (!CryptEncodeObjectEx(X509_ASN_ENCODING, PKCS_PRIVATE_KEY_INFO,
+                           &private_key_info, 0, nullptr, nullptr,
+                           &pkcs8_encoded_length)) {
+    return InvalidArgumentError(
+        FormatWin32Errors("Could not encode private key from PKCS#12 file (",
+                          source, "): "),
+        GCP_ERROR_INFO());
+  }
+  std::vector<BYTE> pkcs8_encoded(pkcs8_encoded_length);
+  CryptEncodeObjectEx(X509_ASN_ENCODING, PKCS_PRIVATE_KEY_INFO,
+                      &private_key_info, 0, nullptr, pkcs8_encoded.data(),
+                      &pkcs8_encoded_length);
+  return pkcs8_encoded;
+}
+
+StatusOr<std::vector<char>> Base64Encode(absl::Span<BYTE const> blob,
+                                         std::string const& source) {
+  DWORD base64_length;
+  if (!CryptBinaryToStringA(blob.data(), static_cast<DWORD>(blob.size()),
+                            CRYPT_STRING_BASE64 | CRYPT_STRING_NOCR, nullptr,
+                            &base64_length)) {
+    return InvalidArgumentError(
+        FormatWin32Errors(
+            "Could not base64 encode private key from PKCS#12 file (", source,
+            "): "),
+        GCP_ERROR_INFO());
+  }
+  std::vector<char> private_key(base64_length);
+  CryptBinaryToStringA(blob.data(), static_cast<DWORD>(blob.size()),
+                       CRYPT_STRING_BASE64 | CRYPT_STRING_NOCR,
+                       private_key.data(), &base64_length);
+  private_key.pop_back();  // remove the null terminator
+  return private_key;
+}
+}  // namespace
```

Comment:
nit: blank line before closing the namespace.

```suggestion
}

}  // namespace
```

Reply:
Done.